### PR TITLE
Rolling window endpoints inclusion

### LINKED ISSUE #13965: Exact matches in time-based .rolling()
I would like to `allow_exact_matches` in time-based `.rolling()`. Using an example similar to the one found in the documentation:

```
df = pd.DataFrame({'B': [1, 1, 2, np.nan, 4]},
                  index = [pd.Timestamp('20130101 09:00:00'),
                           pd.Timestamp('20130101 09:00:02'),
                           pd.Timestamp('20130101 09:00:03'),
                           pd.Timestamp('20130101 09:00:05'),
                           pd.Timestamp('20130101 09:00:06')])
```

I can compute the three-second rolling sum:

```
In [200]: df
Out[200]:
                       B
2013-01-01 09:00:00  1.0
2013-01-01 09:00:02  1.0
2013-01-01 09:00:03  2.0
2013-01-01 09:00:05  NaN
2013-01-01 09:00:06  4.0

In [201]: df.rolling('3s').sum()
Out[201]:
                       B
2013-01-01 09:00:00  1.0
2013-01-01 09:00:02  2.0
2013-01-01 09:00:03  3.0
2013-01-01 09:00:05  2.0
2013-01-01 09:00:06  4.0
```

But note that `09:00:03` does not include `09:00:00`, `09:00:05` does not include `09:00:02`, and `09:00:06` does not include `09:00:03`. If I were to include these timestamps, then I would have:

```
                       B
2013-01-01 09:00:00  1.0
2013-01-01 09:00:02  2.0
2013-01-01 09:00:03  4.0
2013-01-01 09:00:05  3.0
2013-01-01 09:00:06  6.0
```

A quick-and-dirty way of getting these values is with:

```
from itertools import starmap

window = pd.Timedelta('3s')

begins = np.searchsorted(df.index, df.index - window)
ends = np.arange(len(df)) + 1
ranges = starmap(range, zip(begins, ends))

pd.concat([df.ix[r].sum() for r in ranges])
```

`pd.merge_asof()` has the parameter `allow_exact_matches` to permit this. Would it be possible to add this to `.rolling()`?

I can try this one myself.

### PR DESCRIPTION
**Pull request by carlosdanielcsantos**
- [x] closes #13965 
 - [x] time-based window tests added / passed
 - [x] fixed window tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

### CODE REVIEW

File: pandas/core/window.pyx [hunk 1]
```diff
@@ -1141,11 +1181,11 @@ def roll_min(ndarray[numeric] input, int64_t win, int64_t minp,
     index: ndarray, optional
        index for window computation
     """
-    return _roll_min_max(input, win, minp, index, is_max=0)
+    return _roll_min_max(input, win, minp, index, is_max=0, closed=closed)
 
 
 cdef _roll_min_max(ndarray[numeric] input, int64_t win, int64_t minp,
-                   object index, bint is_max):
+                   object index, str closed, bint is_max):
```
Comment (carlosdanielcsantos) [inline pandas/core/window.pyx]: Not sure about the behavior of str here...

Comment (carlosdanielcsantos) [inline pandas/core/window.pyx]: Not sure about the behavior of str here (cdef)

File: pandas/core/window.py [hunk 2]
```diff
@@ -101,6 +103,9 @@ def validate(self):
         if self.min_periods is not None and not \
            is_integer(self.min_periods):
             raise ValueError("min_periods must be an integer")
+        if self.closed not in ['right', 'both', 'left', 'neither']:
```
Comment (jreback) [inline pandas/core/window.py]: testing this?

File: pandas/core/window.pyx [hunk 3]
```diff
@@ -161,7 +161,8 @@ cdef class MockFixedWindowIndexer(WindowIndexer):
 
     """
     def __init__(self, ndarray input, int64_t win, int64_t minp,
-                 object index=None, object floor=None):
+                 object index=None, object floor=None,
+                 bint l_closed=False, bint r_closed=True):
```
Comment (jreback) [inline pandas/core/window.pyx]: if you have a chance, can you add a mini-doc strings to things.

File: pandas/core/window.pyx [hunk 4]
```diff
@@ -307,18 +322,35 @@ def get_window_indexer(input, win, minp, index, floor=None,
         compat Indexer that allows us to use a standard
         code path with all of the indexers.
```
Comment (jreback) [inline pandas/core/window.pyx]: can you augment the doc-string here (with closed)

File: pandas/core/window.pyx [hunk 5]
```diff
@@ -307,18 +322,35 @@ def get_window_indexer(input, win, minp, index, floor=None,
...
     -------
     tuple of 1d int64 ndarrays of the offsets & data about the window
 
     """
 
+    cdef:
+        bint l_closed = False
+        bint r_closed = False
+
+    if closed not in ['right', 'left', 'both', 'neither']:
```
Comment (jreback) [inline pandas/core/window.pyx]: you can just do an assert here (this is an internal routine that has certain guarantees)

File: pandas/core/window.pyx [hunk 6]
```diff
@@ -327,7 +359,7 @@ def get_window_indexer(input, win, minp, index, floor=None,
 
 
 def roll_count(ndarray[double_t] input, int64_t win, int64_t minp,
-               object index):
+               object index, closed='right'):
```
Comment (jreback) [inline pandas/core/window.pyx]: just make this required

``object closed``

File: pandas/core/window.pyx [hunk 7]
```diff
@@ -523,7 +556,7 @@ cdef inline void remove_mean(double val, Py_ssize_t *nobs, double *sum_x,
 
 
 def roll_mean(ndarray[double_t] input, int64_t win, int64_t minp,
-              object index):
+              object index, str closed):
```
Comment (jreback) [inline pandas/core/window.pyx]: ``object closed``

File: pandas/core/window.pyx [hunk 8]
```diff
@@ -1111,7 +1148,7 @@ cdef inline numeric calc_mm(int64_t minp, Py_ssize_t nobs,
 
 
 def roll_max(ndarray[numeric] input, int64_t win, int64_t minp,
-             object index):
+             object index, str closed):
```
Comment (jreback) [inline pandas/core/window.pyx]: same for the rest of these: str-> object

File: pandas/core/window.pyx (hunk 1 above)
Comment (jreback) [inline pandas/core/window.pyx]: same

File: pandas/tests/test_window.py [hunk 9]
```diff
@@ -3356,6 +3356,34 @@ def test_min_periods(self):
         result = df.rolling('2s', min_periods=1).sum()
         tm.assert_frame_equal(result, expected)
 
+    def test_closed(self):
+        df = DataFrame({'A': [1]*5},
```
Comment (jreback) [inline pandas/tests/test_window.py]: add this issue number as a comment

File: pandas/tests/test_window.py [hunk 10]
```diff
@@ -3356,6 +3356,34 @@ def test_min_periods(self):
...
+                        index = [pd.Timestamp('20130101 09:00:01'),
+                                 pd.Timestamp('20130101 09:00:02'),
+                                 pd.Timestamp('20130101 09:00:03'),
```
Comment (jreback) [inline pandas/tests/test_window.py]: test passing invalid closed

File: pandas/tests/test_window.py [hunk 11]
```diff
@@ -3356,6 +3356,34 @@ def test_min_periods(self):
...
+        expected = df.copy()
+        expected["A"] = [1.0, 2, 3, 3, 2]
+        result = df.rolling('2s', closed='both').sum()
+        tm.assert_frame_equal(result, expected)
+
+        expected = df.copy()
+        expected["A"] = [np.nan, 1.0, 2, 2, 1]
+        result = df.rolling('2s', closed='left').sum()
+        tm.assert_frame_equal(result, expected)
+
```
Comment (jreback) [inline pandas/tests/test_window.py]: @chrisaycock how do these tests look? what corners do we need to test here

**Comment by carlosdanielcsantos** [comment]
Here is a quick example of what this feature looks like

```
df = pd.DataFrame({'x': [1]*5},
....:                 index = [pd.Timestamp('20130101 09:00:01'),
....:                          pd.Timestamp('20130101 09:00:02'),
....:                          pd.Timestamp('20130101 09:00:03'),
....:                          pd.Timestamp('20130101 09:00:04'),
....:                          pd.Timestamp('20130101 09:00:06')])

df["right"] = df.rolling('2s', closed='right').x.sum()   # current behavior
df["both"] = df.rolling('2s', closed='both').x.sum()
df["left"] = df.rolling('2s', closed='left').x.sum()
df["open"] = df.rolling('2s', closed='none').x.sum()

print(df)
                     x  right  both  left  open
2013-01-01 09:00:01  1    1.0   1.0   NaN   NaN
2013-01-01 09:00:02  1    2.0   2.0   1.0   1.0
2013-01-01 09:00:03  1    2.0   3.0   2.0   1.0
2013-01-01 09:00:04  1    2.0   3.0   2.0   1.0
2013-01-01 09:00:06  1    1.0   2.0   1.0   NaN
```

**Review by jreback — CHANGES_REQUESTED** [review]
mostly minor points
- add a whatsnew entry (Other Enhancements) in whatsnew (0.20.0), with a link to the computation.rst docs (where you have a new section showing the use of closed; you can use your test examples, e.g. showing all 4 cases).

**Comment by jreback** [comment]
@carlosdanielcsantos https://github.com/pandas-dev/pandas/pull/15795#issuecomment-288883027 are basically the docs (with a bit of prose) and nice formatting. Laying out the possibilities (in a sub-section of computation.rst). Note you also might want to have a 'real-world' uses case sentence for some of these.

**Comment by jreback** [comment]
you can see pep errors: http://pandas-docs.github.io/pandas-docs-travis/contributing.html#python-pep8

use ``git diff master | flake8 --diff``

File: pandas/tests/test_window.py (hunk 11 above)
Comment (chrisaycock) [inline pandas/tests/test_window.py]: What happens if the DataFrame is empty?

Comment (carlosdanielcsantos) [inline pandas/tests/test_window.py]: If the DataFrame is empty, the behavior is exactly the same as in the current implementation. For time-based windows, it throws "ValueError: window must be an integer".  Note that this feature only affects the window indexer's start and end vectors. In fixed windows, empty DataFrames always return empty (the result is constructed as np.empty(N) where N is the length of the input)

Comment (jreback) [inline pandas/tests/test_window.py]: separately I think [4] should just return an empty frame for consistency.

```
In [4]: DataFrame().rolling('1s').sum()
ValueError: window must be an integer

In [5]: DataFrame().rolling(1).sum()
Out[5]: 
Empty DataFrame
Columns: []
Index: []
```

@chrisaycock ?

**Comment by carlosdanielcsantos** [comment]
@jreback How should I add this to computations.rst? A new section inside Window Functions or just add the examples inside the existing Window Functions -> Rolling Windows?

File: doc/source/whatsnew/v0.20.0.txt [hunk 12]
```diff
@@ -268,6 +268,7 @@ To convert a ``SparseDataFrame`` back to sparse SciPy matrix in COO format, you
 Other enhancements
 ^^^^^^^^^^^^^^^^^^
 
+- ``DataFrame.rolling()`` now accepts the parameter ``closed='right'|'left'|'both'|'neither'`` to choose the rolling window endpoint closedness. (:issue:`13965`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: is that a word?

File: pandas/core/window.py [hunk 13]
```diff
@@ -374,8 +379,12 @@ class Window(_Window):
     on : string, optional
         For a DataFrame, column on which to calculate
         the rolling window, rather than the index
+    closed : 'right', 'left', 'both', 'neither'
```
Comment (jreback) [inline pandas/core/window.py]: add the default setting (right)

Comment (jreback) [inline pandas/core/window.py]: I would do this with a bulleted list

File: pandas/core/window.pyx [hunk 14]
```diff
@@ -158,10 +158,15 @@ cdef class MockFixedWindowIndexer(WindowIndexer):
         index of the input
     floor: optional
         unit for flooring
+    l_closed: bint
+        left endpoint closedness
```
Comment (jreback) [inline pandas/core/window.pyx]: I would write these out ``l_closed`` -> ``left_closed`` easier to read

File: pandas/core/window.pyx [hunk 15]
```diff
@@ -229,10 +239,14 @@ cdef class VariableWindowIndexer(WindowIndexer):
         min number of obs in a window to consider non-NaN
     index: ndarray
         index of the input
+    l_closed: bint
+        left endpoint closedness
```
Comment (jreback) [inline pandas/core/window.pyx]: can add a bit of explication here (1 line), just for future readers

File: pandas/core/window.pyx [hunk 16]
```diff
@@ -261,7 +276,11 @@ cdef class VariableWindowIndexer(WindowIndexer):
         N = self.N
 
         start[0] = 0
-        end[0] = 1
+
```
Comment (jreback) [inline pandas/core/window.pyx]: put the comments above the code

```
if r_closed:
    # .....
    end[0] = 1
else:
    # ....
    end[0]  0
```

File: pandas/core/window.pyx [hunk 17]
```diff
@@ -271,6 +290,9 @@ cdef class VariableWindowIndexer(WindowIndexer):
                 end_bound = index[i]
                 start_bound = index[i] - win
 
+                if l_closed:  # left endpoint is closed
```
Comment (jreback) [inline pandas/core/window.pyx]: same

Comment (jreback) [inline pandas/core/window.pyx]: of can do (I use this if only an if statement)
```
# left endpoint ....
if left_closed:
    ....
```

File: pandas/core/window.pyx [hunk 18]
```diff
@@ -286,9 +308,13 @@ cdef class VariableWindowIndexer(WindowIndexer):
                 else:
                     end[i] = end[i - 1]
 
+                # right endpoint is open
```
Comment (jreback) [inline pandas/core/window.pyx]: yep excatly like this one

File: pandas/core/window.pyx [hunk 19]
```diff
@@ -299,6 +325,8 @@ def get_window_indexer(input, win, minp, index, floor=None,
     minp: integer, minimum periods
     index: 1d ndarray, optional
         index to the input array
+    closed: 'right', 'left', 'both', 'neither'
```
Comment (jreback) [inline pandas/core/window.pyx]:
```
closed: string, default 'right'
        {'right',......}
```

**Comment by jreback** [comment]
@carlosdanielcsantos lgtm. some more doc-string nitpicks.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
ping when all green.

**Comment by carlosdanielcsantos** [comment]
@jreback I was going to start writing tests for the FixedWindowIndexer but there are no explicit rolling computing test going on in TestRolling (they are done in TestMoments, I think). Where should I add these new tests?

File: doc/source/whatsnew/v0.20.0.txt (hunk 12 above)
Comment (carlosdanielcsantos) [inline doc/source/whatsnew/v0.20.0.txt]: There is some discussion going on about that online, but its a useful word to reference the property of being closed. But I am open to suggestions, no pun intended :)

**Comment by jreback** [comment]
> @jreback How should I add this to computations.rst? A new section inside Window Functions or just add the examples inside the existing Window Functions -> Rolling Windows?

exactly, a new sub-section that pretty much shows the example with all 4 closed. maybe right before Centering Windows?

**Comment by jreback** [comment]
> @jreback I was going to start writing tests for the FixedWindowIndexer but there are no explicit rolling computing test going on in TestRolling (they are done in TestMoments, I think). Where should I add these new tests?

We don't have explicit tests for the cython routines, these are covered by the integration tests (top-level routines). Though certainly could add them if you want / can. 

You could add a new class in test_windows that explicity exercises these.

File: doc/source/whatsnew/v0.20.0.txt (hunk 12 above)
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: hahah yep that is fine

Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: maybe ``closed-ness``? (though I don't think *that* is actually a word)

Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: also add a ``:ref:`` to the new sub-section that you are creating in computation.rst

**Comment by carlosdanielcsantos** [comment]
@jreback @chrisaycock 
Thinking deeper about the fixed window case, it seems that, intuitively, the default behavior right now better resembles the closed='both' case. Notice that when you use rolling(3).apply(f), the result for index i will be the result of f applied to `[i-2, i-1, i]`, not `]i-2, i-1, i]` (ISO 31-11 math set notation). One could argue that this is equivalent to having closed='right' with f applied to `]i-3, i-2, i-1, i]`. Indeed, but this probably adulterates the meaning of the window size.

**Comment by jreback** [comment]
> @jreback @chrisaycock
Thinking deeper about the fixed window case, it seems that, intuitively, the default behavior right now better resembles the closed='both' case. Notice that when you use rolling(3).apply(f), the result for index i will be the result of f applied to [i-2, i-1, i], not ]i-2, i-1, i] (ISO 31-11 math set notation). One could argue that this is equivalent to having closed='right' with f applied to ]i-3, i-2, i-1, i]. Indeed, but this probably adulterates the meaning of the window size.

@carlosdanielcsantos can you show a comparison of these 2, to show more what you mean?

**Comment by jreback** [comment]
> Nonetheless, I think that the possibility of ignoring the current index in computations is valuable (that was my initial problem, after all). Some solutions for this are:

This a bit orthogonal though. We *always* want to actually use the index. Since using a fixed window size, it *looks* like we are ignoring the index :>.

So a possiblity is to actually check the index whether its equiv to a RangeIndex. e.g.

```
In [5]: df1 = pd.DataFrame([1] * 5)

In [6]: df1.rolling(3, min_periods=1).sum()
Out[6]: 
     0
0  1.0
1  2.0
2  3.0
3  3.0
4  3.0

In [8]: df2 = pd.DataFrame([1] * 5, index=[0, 1, 3, 4, 5])

In [9]: df2.rolling(3, min_periods=1).sum()
Out[9]: 
     0
0  1.0
1  2.0
3  3.0
4  3.0
5  3.0
```

```
In [11]: (np.diff(df1.index)==1).all()
Out[11]: True

In [12]: (np.diff(df2.index)==1).all()
Out[12]: False

```

Let's deal with this in a separate issue. 

A way to disambiguate on this ATM is to allow ``closed=None`` to be the default parameter. Then it cannot be anything but that for a FixedWindow. For Variable, if we see ``closed=None`` -> ``closed='right'`` IOW the default is the same. And then allow the rest of the closed intervals.

I think *we* should fix this for Fixed intervals, but I agree this is a problem because we essentially have ``closed='both'`` as the default now (its implicit though) (as you stated above).

**Comment by jreback** [comment]
@carlosdanielcsantos if you can update would be great (docs and such), ignoring the issue about fixed windows for the moment.

**Comment by carlosdanielcsantos** [comment]
I'll do that in the next hours. Should I just leave the fixed-window as it is?

**Comment by jreback** [comment]
I think make the ``closed=None`` the default, and document the possibilities (and default) for the variable sized window (and validate left,right,both,neither). For the fixed window document as well (you can say its closed='both', but ATM only allow None). and then open a new issue where we can deal with this.

We have to fix this, but this PR merely opens up the possibilities of change here. So I think ok to merge this.

**Comment by jreback** [comment]
@carlosdanielcsantos can you update

**Comment by carlosdanielcsantos** [comment]
@jreback made some corrections but circleci is not passing in test_append_index - pandas.tests.test_multilevel.TestMultiLevel, but passing locally. Any thoughts?

**Comment by jreback** [comment]
try pushing again (make a modification)
looks like a fluke

**Comment by carlosdanielcsantos** [comment]
@jreback Still not passing. Could I have broken this with the changes in window?

**Comment by jreback** [comment]
@carlosdanielcsantos I rebased you against current master. This should pass cleanly. If you'd make those other changes and ping.

**Comment by carlosdanielcsantos** [comment]
@jreback what changes? I believe I updated everything that you requested. Can you take a look at AppVeyor build that failed?

File: doc/source/computation.rst [hunk 20]
```diff
@@ -459,6 +459,43 @@ default of the index) in a DataFrame.
    dft
    dft.rolling('2s', on='foo').sum()
 
+.. _stats.rolling_window.endpoints:
+
+Rolling window endpoint inclusion
+~~~~~~~~~~~~~~~~~~
```
Comment (jreback) [inline doc/source/computation.rst]: make this Rolling Window Endpoints

File: doc/source/computation.rst [hunk 21]
```diff
@@ -459,6 +459,43 @@ default of the index) in a DataFrame.
...
+
+.. versionadded:: 0.20.0
+
+The inclusion of the interval endpoints in rolling window calculations can be specified with the ``closed``
+parameter:
+
+  - ``right`` : close right endpoint (default for time-based windows)
```
Comment (jreback) [inline doc/source/computation.rst]: can you make this a csv-table (3 columns: ``closed``, ``description``, ``default``) or something like that

File: doc/source/computation.rst [hunk 22]
```diff
@@ -459,6 +459,43 @@ default of the index) in a DataFrame.
...
+                              pd.Timestamp('20130101 09:00:06')])
+
+   df["right"] = df.rolling('2s', closed='right').x.sum()  # default
+   df["both"] = df.rolling('2s', closed='both').x.sum()
+   df["left"] = df.rolling('2s', closed='left').x.sum()
+   df["neither"] = df.rolling('2s', closed='neither').x.sum()
+
+   df
+
+Currently, this feature is only implemented for time-based windows.
```
Comment (jreback) [inline doc/source/computation.rst]: expand this a touch to say that *only* ``closed='both'`` is accepted for fixed windows.

File: doc/source/whatsnew/v0.20.0.txt [hunk 23]
```diff
@@ -319,6 +319,7 @@ To convert a ``SparseDataFrame`` back to sparse SciPy matrix in COO format, you
 Other Enhancements
 ^^^^^^^^^^^^^^^^^^
 
+- ``DataFrame.rolling()`` now accepts the parameter ``closed='right'|'left'|'both'|'neither'`` to choose the rolling window endpoint closedness. See :ref:`documentation <stats.rolling_window.endpoints>` (:issue:`13965`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: See ``:ref:`the documentation ....``

File: pandas/core/window.py [hunk 24]
```diff
@@ -101,6 +103,10 @@ def validate(self):
         if self.min_periods is not None and not \
            is_integer(self.min_periods):
             raise ValueError("min_periods must be an integer")
+        if self.closed is not None and self.closed not in \
+           ['right', 'both', 'left', 'neither']:
+            raise ValueError("closed must be 'right', 'left', 'both' or "
```
Comment (jreback) [inline pandas/core/window.py]: is there validation for a fixed window when closed is NOT both?

**Review by jreback — APPROVED** [review]
minor comments.

File: doc/source/computation.rst (hunk 22 above)
Comment (carlosdanielcsantos) [inline doc/source/computation.rst]: I am only accepting closed=None for fixed windows, as previously requested by you

> For the fixed window document as well (you can say its closed='both', but ATM only allow None)

Comment (jreback) [inline doc/source/computation.rst]: that's what i mean just make sure the docs are clear on that ; e.g. None -> 'both' for fixed (and assert that as well )

File: pandas/core/window.py (hunk 24 above)
Comment (carlosdanielcsantos) [inline pandas/core/window.py]: Again, there is a validation for when closed is not None in Rolling.validate()

File: pandas/tests/test_window.py [hunk 25]
```diff
@@ -431,6 +431,12 @@ def test_numpy_compat(self):
             tm.assertRaisesRegexp(UnsupportedFunctionCall, msg,
                                   getattr(r, func), dtype=np.float64)
 
+    def test_closed(self):
```
Comment (carlosdanielcsantos) [inline pandas/tests/test_window.py]: @jreback here is the piece that tests that closed must be None for fixed-window. Well, actually it only tests that it should not be 'neither', but I do not think replication is necessary.

File: pandas/core/window.py [hunk 26]
```diff
@@ -1044,6 +1058,10 @@ def validate(self):
         elif self.window < 0:
             raise ValueError("window must be non-negative")
 
+        if not self.is_datetimelike and self.closed is not None:
```
Comment (carlosdanielcsantos) [inline pandas/core/window.py]: @jreback here we enforce that `closed` must be `None` for windows that are not datetimelike

**Comment by jreback** [comment]
thanks @carlosdanielcsantos nice PR. as you can see, the code changes are the 'easy' part, its the docs & testing that are the details.!

keem em coming!

**Pull request closed without merging**